### ISSUE TITLE
Multiple one-to-one relations in one entity generate differently in migration

### ISSUE THREAD

**Issue by t-mxcom** [dotnet/efcore#27904]
## Multiple one-to-one relations in one entity generate differently in migration - but why?

Good morning (it's about 9am here in Austria)!

I want to add two one-to-one (one-to-zero or one) relations to my parent entity (=`UserEntity`).
Let me express this in code.

```cs
public class UserEntity
{
    // Columns of the UserEntity
    public Guid Id { get; set; }
    public byte[] RowVersion { get; set; }
    public string Name { get; set; }

    // Navigation Properties
    public virtual MfaSettingsEntity MfaSettings { get; set; }
    public virtual AuthenticatorSettingsEntity AuthenticatorSettings { get; set; }
}

public class MfaSettingsEntity
{
    // Columns of the MfaSettingsEntity
    public Guid Id { get; set; }
    public byte[] RowVersion { get; set; }
    public Guid UserId { get; set; } // this is the key reference to UserEntity
    public bool MfaEnabled { get; set; }

    // no Navigation Properties here
}

public class AuthenticatorSettingsEntity
{
    // Columns of the AuthenticatorSettingsEntity
    public Guid Id { get; set; }
    public byte[] RowVersion { get; set; }
    public Guid UserId { get; set; } // this is the key reference to UserEntity
    public string AuthenticatorKey { get; set; }

    // no Navigation Properties here
}
```

I'm using `EntityTypeConfiguration` classes to initialize the model.

```cs
public class UserEntityTypeConfiguration : IEntityTypeConfiguration<UserEntity>
{
    public void Configure(EntityTypeBuilder<UserEntity> builder)
    {
        builder.ToTable("Users", "dbo");
        builder.HasKey(property => property.Id).IsClustered(false);

        builder.Property(property => property.Id).ValueGeneratedOnAdd().IsRequired(true);
        builder.Property(property => property.RowVersion).IsRowVersion();
        builder.Property(property => property.Name).HasMaxLength(100).IsRequired(true);

        builder
            .HasOne<MfaSettingsEntity>(entity => entity.MfaSettings)
            .WithOne()
            .HasForeignKey<MfaSettingsEntity>(entity => entity.UserId);

        builder
            .HasOne<AuthenticatorSettingsEntity>(entity => entity.AuthenticatorSettings)
            .WithOne()
            .HasForeignKey<AuthenticatorSettingsEntity>(entity => entity.UserId);
    }
}

public class MfaSettingsEntityTypeConfiguration : IEntityTypeConfiguration<MfaSettingsEntity>
{
    public void Configure(EntityTypeBuilder<MfaSettingsEntity> builder)
    {
        builder.ToTable("MfaSettings", "dbo");
        builder.HasKey(property => property.Id).IsClustered(false);

        builder.Property(property => property.Id).ValueGeneratedOnAdd().IsRequired(true);
        builder.Property(property => property.RowVersion).IsRowVersion();
        builder.Property(property => property.UserId).IsRequired(true);
        builder.Property(property => property.MfaEnabled);
    }
}

public class AuthenticatorSettingsEntityTypeConfiguration : IEntityTypeConfiguration<AuthenticatorSettingsEntity>
{
    public void Configure(EntityTypeBuilder<AuthenticatorSettingsEntity> builder)
    {
        builder.ToTable("AuthenticatorSettings", "dbo");
        builder.HasKey(property => property.Id).IsClustered(false);

        builder.Property(property => property.Id).ValueGeneratedOnAdd().IsRequired(true);
        builder.Property(property => property.RowVersion).IsRowVersion();
        builder.Property(property => property.UserId).IsRequired(true);
        builder.Property(property => property.AuthenticatorKey).HasMaxLength(100).IsRequired(true);
    }
}
```

This definition results in the following migration being created.

*Notes:*

- The `Users` table resulting from the `UserEntity` was created by a preceding migration.  
  Thus the following migration just contains the differences.
- The only modification to the `UserEntity` was adding the navigation properties.
- I removed the `Down` method to shorten the code in this post.
- I know that I could simlpy add the columns of the child entities to the parent entity, but this is something I don't want to do.

Please note the comments prefixed with `QUESTION` and `EXPECTATION` that I added!

```cs
public partial class MyMigration : Migration
{
    protected override void Up(MigrationBuilder migrationBuilder)
    {
        // QUESTION: Why is this column added and why is it only done for AuthenticatorSettings, but not for MfaSettings?
        // EXPECTATION: I don't want this column to be created!
        migrationBuilder.AddColumn<Guid>(
            name: "AuthenticatorSettingsId",
            schema: "dbo",
            table: "Users",
            nullable: true);

        migrationBuilder.CreateTable(
            name: "AuthenticatorSettings",
            schema: "dbo",
            columns: table => new
            {
                Id = table.Column<Guid>(nullable: false),
                RowVersion = table.Column<byte[]>(rowVersion: true, nullable: true),
                UserId = table.Column<Guid>(nullable: false),
                AuthenticatorKey = table.Column<string>(maxLength: 100, nullable: false)
            },
            constraints: table =>
            {
                table.PrimaryKey("PK_AuthenticatorSettings", x => x.Id)
                    .Annotation("SqlServer:Clustered", false);
                table.ForeignKey(
                    name: "FK_AuthenticatorSettings_Users_UserId",
                    column: x => x.UserId,
                    principalSchema: "dbo",
                    principalTable: "Users",
                    principalColumn: "Id",
                    onDelete: ReferentialAction.Cascade);
            });

        migrationBuilder.CreateTable(
            name: "MfaSettings",
            schema: "dbo",
            columns: table => new
            {
                Id = table.Column<Guid>(nullable: false),
                RowVersion = table.Column<byte[]>(rowVersion: true, nullable: true),
                UserId = table.Column<Guid>(nullable: false),
                MfaEnabled = table.Column<bool>(nullable: false)
            },
            constraints: table =>
            {
                table.PrimaryKey("PK_MfaSettings", x => x.Id)
                    .Annotation("SqlServer:Clustered", false);
                table.ForeignKey(
                    name: "FK_MfaSettings_Users_UserId",
                    column: x => x.UserId,
                    principalSchema: "dbo",
                    principalTable: "Users",
                    principalColumn: "Id",
                    onDelete: ReferentialAction.Cascade);
            });

        // EXPECTATION: If AuthenticatorSettingsId wasn't be created, this index wouldn't be either.
        migrationBuilder.CreateIndex(
            name: "IX_Users_AuthenticatorSettingsId",
            schema: "dbo",
            table: "Users",
            column: "AuthenticatorSettingsId");

        // QUESTION: It's perfect that EF creates this index, but why is this one not marked as "unique" while the following one is?
        // EXPECTATION: I want this index to be unique!
        migrationBuilder.CreateIndex(
            name: "IX_AuthenticatorSettings_UserId",
            schema: "dbo",
            table: "AuthenticatorSettings",
            column: "UserId");

        migrationBuilder.CreateIndex(
            name: "IX_MfaSettings_UserId",
            schema: "dbo",
            table: "MfaSettings",
            column: "UserId",
            unique: true);

        // EXPECTATION: I don't want this foreign key to be created!
        migrationBuilder.AddForeignKey(
            name: "FK_Users_AuthenticatorSettings_AuthenticatorSettingsId",
            schema: "dbo",
            table: "Users",
            column: "AuthenticatorSettingsId",
            principalSchema: "dbo",
            principalTable: "AuthenticatorSettings",
            principalColumn: "Id",
            onDelete: ReferentialAction.Restrict);
    }
}
```

Please help me to understand why EF creates the migration like that and I appreciate your hints on how to solve this!
Thank you very much!

Max

### Provider and version information

EF Core version: 3.1.21
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1, .NET Standard 2.1
Operating system: Microsoft Windows 10
IDE: Microsoft Visual Studio 2022

**Comment by t-mxcom** [comment]
## Update

I changed the navigation properties on `UserEntity` to `ICollection<T>` for both, `MfaSettingsEntity` and `AuthenticatorSettingsEntity`, changed the relations in the `UserEntityTypeConfiguration` class to `HasMany(...).WithOne()` and re-generated the migration.

For `MfaSettings` everything was fine, but `AuthenticatorSettings` again resulted in a weird migration.

```cs
migrationBuilder.CreateTable(
    name: "UserAuthenticatorSettings",
    schema: "Identity",
    columns: table => new
    {
        Id = table.Column<Guid>(nullable: false),
        RowVersion = table.Column<byte[]>(rowVersion: true, nullable: true),
        UserId = table.Column<Guid>(nullable: false),
        AuthenticatorKey = table.Column<string>(type: "varchar(100)", maxLength: 100, nullable: false),
        Registration = table.Column<DateTime>(nullable: true),
        Comment = table.Column<string>(maxLength: 200, nullable: true),
        UserEntityId = table.Column<Guid>(nullable: true)
    },
    constraints: table =>
    {
        table.PrimaryKey("PK_UserAuthenticatorSettings", x => x.Id)
            .Annotation("SqlServer:Clustered", false);
        table.ForeignKey(
            name: "FK_UserAuthenticatorSettings_Users_UserEntityId",
            column: x => x.UserEntityId,
            principalSchema: "Identity",
            principalTable: "Users",
            principalColumn: "Id",
            onDelete: ReferentialAction.Restrict);
        table.ForeignKey(
            name: "FK_UserAuthenticatorSettings_Users_UserId",
            column: x => x.UserId,
            principalSchema: "Identity",
            principalTable: "Users",
            principalColumn: "Id",
            onDelete: ReferentialAction.Cascade);
    });
```

An additional `UserEntityId` column was added to the table definition and again it creates two foreign keys!?!?
One with `Cascade` and one with `Restricted`.

Max

**Comment by ajcvickers** [comment]
@t-mxcom Please try using EF Core 6.0.4, since this may be a bug that has been fixed since the 3.1 release. If the issue still occurs, then please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by t-mxcom** [comment]
@ajcvickers Unfortunately, EF Core 6.0.4 also creates the additional (unexpected) foreign key reference.
I'm now going to extract the necessary code and come back with a runnable small project.
